fathom/src/surface/elaboration.rs

```diff
@@ -1581,39 +1651,67 @@ impl<'interner, 'arena> Context<'interner, 'arena> {
     fn check_fun_lit(
         &mut self,
         range: ByteRange,
-        patterns: &[(Pattern<ByteRange>, Option<&Term<'_, ByteRange>>)],
+        params: &[FunParam<'_, ByteRange>],
         body_expr: &Term<'_, ByteRange>,
         expected_type: &ArcValue<'arena>,
     ) -> core::Term<'arena> {
-        match patterns.split_first() {
-            Some(((pattern, r#type), next_patterns)) => {
+        match params.split_first() {
+            Some((param, next_params)) => {
                 let body_type = self.elim_env().force(expected_type);
                 match body_type.as_ref() {
-                    Value::FunType(_, param_type, next_body_type) => {
-                        let range = ByteRange::merge(&pattern.range(), &body_expr.range()).unwrap();
-                        let pattern = self.check_ann_pattern(pattern, *r#type, param_type);
+                    Value::FunType(param_plicity, _, param_type, next_body_type)
+                        if param.plicity == *param_plicity =>
+                    {
+                        let range =
+                            ByteRange::merge(&param.pattern.range(), &body_expr.range()).unwrap();
+                        let pattern = self.check_ann_pattern(
+                            &param.pattern,
+                            param.r#type.as_ref(),
+                            param_type,
+                        );
                         let (name, arg_expr) = self.push_local_param(pattern, param_type.clone());
 
                         let body_type = self.elim_env().apply_closure(next_body_type, arg_expr);
                         let body_expr =
-                            self.check_fun_lit(range, next_patterns, body_expr, &body_type);
+                            self.check_fun_lit(range, next_params, body_expr, &body_type);
                         self.local_env.pop();
 
-                        core::Term::FunLit(range.into(), name, self.scope.to_scope(body_expr))
+                        core::Term::FunLit(
+                            range.into(),
+                            param.plicity,
+                            name,
+                            self.scope.to_scope(body_expr),
+                        )
+                    }
+                    // If an implicit function is expected, try to generalise the function literal by wrapping it in an implicit function
```

Comment:
Nit: comments are getting a bit long here - I think I generally tend to keep them around 80 characters

Reply:
I think we can enable automatic comment wrapping in `rustfmt.toml`?

Reply:
Nightly only, sadly 😢